### DISCUSSION TITLE
The order of execution

### DISCUSSION THREAD

**Discussion by krismc** [micropython/micropython#16292] [Using MicroPython] [answered]
I sometimes have a guess at the sequence in which upython does things and come up wrong. In the examples below boot_modem & _attached() are functions that return 'OK' or None.

If I run
```
r=0
r=boot_modem() and _attached()
if r and 'OK' in r:
```
all is good. But my original idea
```
r=0
if r:=(boot_modem() and attached()) and 'OK' in r:
```
crashes saying integers aren't iterable, which means the walrus hasn't run before the 'OK' in r check & r must still be zero. I don't understand why that is?

**Comment by GitHubsSilverBullet** [comment]
The ›new‹ r hasn't been defined at this point, the ›old‹ r is still an integer 0.
Thus you're testing whether 'OK' is in 0. 
Try:
```py
    if (r := (boot_modem() and attached())) and 'OK' in r:
```
and all is well.

**Comment by peterhinch** [comment, marked as answer]
Yes. It's usually wise to put assignment expressions in parens. The following works as expected:
```py
>>> r = 0
>>> if (r:= 1) and r==1: print(f"r = {r}")
... 
r = 1
```
But this does not:
```py
>>> r = 0
>>> if r:= 1 and r==1: print(f"r = {r}")
... 
>>> r
False
>>> 
```
because the latter is effectively doing
```py
if r:= (1 and r==1): print(f"r = {r}
```
which sets r `False` (because the comparison returns `False`).

**Reply by krismc** [reply]
That makes sense Peter, tnx. Now that I understand no need to add it to my list of python mongrel.
